### ISSUE TITLE
DJL on Arm

### ISSUE THREAD

**Issue by willishf** [awslabs/djl#195]
## Question
I knew if I waited long enough someone would come up with a Java API/framework to leverage latest and greatest without needing to use python. Thank you!!!

Has any work been done or testing on arm platforms like Jetson Nano, Google Coral or Raspberry PI. Primary use case is running models on the edge and leveraging GPU on Nano, Tensor Flow Light on Coral and CPU on PI. 

Going to assume this is JNI under the covers linking to appropriate C libraries and thus a compile/install package needed for each OS/hardware combination.

**Comment by stu1130** [comment]
Thanks! I have built the PyTorch native code including libtorch & JNI on a1 instance (arm) and it works without any problem. I think it should be able to extend to any arm-based edge device like Jetson Nano. I'll switch the labels to "feature request" as we should have tutorial or example of how to deploy the model on edge devices.

**Comment by lanking520** [comment]
Or probably to epand the use cases we may need to consider adding TensorFlow Lite into DJL's family

**Comment by stu1130** [comment]
Hi @willishf I am going to close the issue in favor of #375

**Comment by willishf** [comment]
I have made reasonably good process on leveraging gstreamer pipeline for
image processing on jetson nano in Java. Nvidia is making a big investment
in DeepStream which is based on gstreamer to stay in gpu pipeline as long
as possible for high FPS from megapixel camera.

Having mxnet model conversion support that runs in gstreamer would solve
many problems related to megapixel camera access and high FPS. DeepStream
is closed source.

DJL via gstreamer would have full control of pipeline running on jetson
nano hardware that is gpu enabled. To run on raspberry pi the gstreamer
pipeline would not have native gpu image conversion tools but would be a
gstreamer raspberry pi equivalent pipeline.

Can do a demo if that would help fill in the details.

**Comment by stu1130** [comment]
Hi @willishf that's awesome!

What do you mean by `Having mxnet model conversion support`? If the goal is to run MXNet model on Jetson, I think MXNet python can already do that. We just need to include their native libraries(libmxnet.so) into DJL. Or does it require other integration with gstreamer?

**Comment by willishf** [comment]
It is my understanding that one of the goals of DJL is being model agnostic
and hiding those implementation details. That was the point I was trying to
make. Having a gstreamer plug in that runs inference on GPU regardless of
framework would be a very nice feature. On embedded systems with hardware
pipelines from camera to GPU you want to avoid touching the image data at
the CPU level for optimal performance. Gstreamer is a C api/library focused
on setting up pipelines that have inputs and outputs. If you can make the
inputs match the outputs and the outputs match the inputs then you can
basically write a pipeline and hand it over to gstreamer and it makes it
happen. You can do just about anything with gstreamer related to video and
audio.

The following pipeline is what I launch in Java to tell gstreamer I want
the NVIDIA supplied camera interface application *nvarguscamerasrc *to
connect to the camera with attributes of memory type, resolution and
format. The second step is *queue *that the camera only supports 30fps but
I can only process 5 fps so don't buffer and drop the oldest frames because
I only want to see the most recent image. The *nvvidconv *application runs
on the GPU and can take images directly from camera and is the swiss army
knife of format, size, cropping, resolution, rotation all done on GPU.
The *identity
*step is where I hook into the pipeline and register a method to be called
for each new image, I process the image and then put it back in the
pipeline for the next step. The *nveglglessink *step runs GPU application
provided by NVIDIA to display the image. I could add in a h264. h265, V8
etc encoding step that if GPU aware would run on GPU. You can then add a
h264 payload wrapper step and then write to storage or stream on the
internet.


gstreamerPipeLine = "nvarguscamerasrc ! video/x-raw(memory:NVMM),
width=1640, height=1232, format=NV12 ! queue max-size-buffers=1
leaky=downstream ! nvvidconv ! video/x-raw, width=1640, height=1232,
format=RGBA ! identity name=identity ! nveglglessink -e"

From a deep learning perspective and what DeepStream does is to add an
inference engine with a specific supported model and run on GPU directly.
If all of your steps run on GPU you can take advantage of the hardware
pipeline for performance. I would then hook the Java application into the
step after the gstreamer inference engine and consume meta data via call
back and only touch the image data if needed. Gstreamer is also supported
on Google Coral but using google supplied applications that know how to
leverage the hardware pipeline. This also works on Linux with the latest
and greatest NVIDIA GPU cards and windows as well. The gstreamer pipeline
has a large number of tools in the tool box that run on windows and Linux
usually by the same attribute description. NVIDIA provides tools that will
only work on Linux as an example.  Using the GPU in Gstreamer for real time
h264/h265 encoding has been around for awhile. Using gstreamer to do
inference on GPU is the way to go for performance.
